peps/pep-9999.rst

```diff
@@ -0,0 +1,120 @@
+Python-Version: 3.15
+Post-History: 02-Jun-2019,
+              `09-May-2025 <https://discuss.python.org/t/91337>`__,
+
+
+Abstract
+========
+
+This PEP describes new module for number-theoretic functions, like
+:external+py3.14:func:`math.gcd` or :external+py3.14:func:`math.isqrt`.
+
+
+Motivation
+==========
+
+The :external+py3.14:mod:`math` documentaton says: "This module provides
+access to the mathematical functions defined by the C standard."  But as a
+state of art, over time the module was populated with functions that aren't
+related to the C standard or floating-point arithmetics.  Now it's much
+harder to describe module scope, content and interfaces (e.g. returned
+values).
+
+Apparently, the :external+py3.14:mod:`math` can't serve as a catch-all place
+for mathematical functions: we have also the :external+py3.14:mod:`cmath` and
+the :external+py3.14:mod:`statistics`.  Lets make same for integer-related
+functions.
+
+
+Specification
+=============
+
+The PEP proposes moving the following integer-related functions [1]_ to a new
+module, called ``imath``:
+
+* :external+py3.14:func:`~math.gcd`
+* :external+py3.14:func:`~math.lcm`
+* :external+py3.14:func:`~math.isqrt`
+* :external+py3.14:func:`~math.factorial`
+* :external+py3.14:func:`~math.comb`
+
+Their aliases in :external+py3.14:mod:`math` will be :term:`soft deprecated`.
+
+Possible extensions for the new module and it's scope are discussed in the
+Open Issues section.
+
+
+Backwards Compatibility
+=======================
+
+As aliases in :external+py3.14:mod:`math` will be kept for indefinite
+time (their use would be discouraged), there are no anticipated code breaks.
+
+
+Reference Implementation
+========================
+
+TBD
+
+
+Open Issues
+===========
+
+Module name
+-----------
+
+Another variants for name: ``ntheory`` (like SymPy's submodule), ``intmath``,
+``integermath`` and ``imaths``.
+
+Possible renaming of isqrt
+--------------------------
+
+As there is no clash with floating-point version, it could be renamed to
+``sqrt`` (Java-like).
+
+Module scope and possible extensions
+------------------------------------
+
+Unless we can just provide bindings to some well supported mathematical
+library like the GMP, the module scope should be limited.  For example, no
+primality testing and factorization.
+
+Among proposed in the initial discussion thread (see also [5]_):
+
+* ``c_div()`` --- for integer ceiling divide, see [2]_, [3]_.
+* ``gcdext()`` --- to solve linear Diophantine equation in two variables (the
+  :external+py3.14:class:`int` implementation actually include extended
+  Euclidean algorithm)
+* ``isqrt_rem()`` --- to return both integer square root and a remainder (if
+  integer isn't a perfect square)
+* ``ilog()`` --- integer logarithm, currently :external+py3.14:func:`math.log`
+  has a special handling for integer arguments.  It's unique (wrt other module
+  functions) and not documented so far, see [4]_
+* ``fibonacci()``.
```

Comment:
ilog() and fibonacci() sound reasonable additions.

Reply:
This set of functions (except for ilog()) was required to support mpmath/diofant on top of the [gmp](https://github.com/diofant/python-gmp) as the gmpy2 alternative.

Reply:
In the abstract, you're setting the scope as number theory.  I don't think primality testing should necessarily be added, but it is part of number theory.  So, is your intention to narrow the scope in the abstract?

Reply:
> So, is your intention to narrow the scope in the abstract?

Sure.  The stdlib can't include all pypi.org.

Reply:
I would prefer imath to be as small as possible.